agent/app/agent.go

```diff
@@ -0,0 +1,404 @@
+	if cfg.AcceptInsecureCert {
+		log.Warn("SSL certificate verification disabled. This is not recommended.")
+	}
+	log.Debugf("Loaded config: %s", cfg.String())
+
+	dockerClient, err := engine.NewDockerGoClient(dockerclient.NewFactory(cfg.DockerEndpoint), cfg)
+	if err != nil {
+		// This is also non terminal in the current config
+		log.Criticalf("Error creating Docker client: %v", err)
+		return nil, err
+	}
+
+	return &ecsAgent{
+		ctx:               ctx,
+		ec2MetadataClient: ec2MetadataClient,
+		cfg:               cfg,
+		dockerClient:      dockerClient,
+		// We instantiate our own credentialProvider for use in acs/tcs. This tries
+		// to mimic roughly the way it's instantiated by the SDK for a default
+		// session.
+		credentialProvider: defaults.CredChain(defaults.Config(), defaults.Handlers()),
+	}, nil
+
+}
+
+func (agent *ecsAgent) printVersion() int {
+	version.PrintVersion(agent.dockerClient)
+	return exitcodes.ExitSuccess
+}
+
+func (agent *ecsAgent) start() int {
+	sighandlers.StartDebugHandler()
+
+	containerChangeEventStream := eventstream.NewEventStream(containerChangeEventStreamName, agent.ctx)
+	credentialsManager := credentials.NewManager()
+	state := dockerstate.NewTaskEngineState()
+	imageManager := engine.NewImageManager(agent.cfg, agent.dockerClient, state)
+
+	return agent.doStart(containerChangeEventStream, credentialsManager, state, imageManager)
+}
+
+func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStream,
+	credentialsManager credentials.Manager,
+	state dockerstate.TaskEngineState,
+	imageManager engine.ImageManager) int {
+
+	taskEngine, currentEC2InstanceID, err := agent.newTaskEngine(containerChangeEventStream,
+		credentialsManager, state, imageManager)
+	if err != nil {
+		return exitcodes.ExitTerminal
+	}
+
+	stateManager, err := agent.newStateManager(taskEngine,
+		&agent.cfg.Cluster, &agent.containerInstanceArn, &currentEC2InstanceID)
+	if err != nil {
+		log.Criticalf("Error creating state manager: %v", err)
+		return exitcodes.ExitTerminal
+	}
+
+	client := ecsclient.NewECSClient(agent.credentialProvider, agent.cfg, agent.ec2MetadataClient)
+	err = agent.registerContainerInstance(taskEngine, stateManager, client)
+	if err != nil {
+		if isNonTerminal(err) {
+			return exitcodes.ExitError
+		}
+		return exitcodes.ExitTerminal
+	}
+
+	// Begin listening to the docker daemon and saving changes
+	taskEngine.SetSaver(stateManager)
+	imageManager.SetSaver(stateManager)
+	taskEngine.MustInit()
+
+	deregisterInstanceEventStream := eventstream.NewEventStream(
+		deregisterContainerInstanceEventStreamName, agent.ctx)
+	deregisterInstanceEventStream.StartListening()
+
+	taskHandler := eventhandler.NewTaskHandler()
+
+	agent.startAsyncRoutines(containerChangeEventStream, credentialsManager, imageManager,
+		taskEngine, stateManager, deregisterInstanceEventStream, client, taskHandler)
+	return agent.startACSSession(credentialsManager, taskEngine, stateManager,
```

Comment:
If the acs handler stops for whatever reason, we want the Agent to exit. We can start it async, add a wait group or something and wait using that. Instead, we chose to do this.